org-code-javabuilder/protocol/src/main/java/org/code/protocol/LifecycleNotifier.java

```diff
@@ -0,0 +1,25 @@
+package org.code.protocol;
+
+import java.util.HashSet;
+import java.util.Set;
+
+/** Notifies delegates of code execution lifecycle events. */
+public class LifecycleNotifier implements LifecycleListener {
```

Comment:
Is it possible to pull the lifecycle changes into a separate PR? It's a pivotal pattern that will be adopted by Javabuilder going forward, and it's worth setting up as a separate PR that establishes the pattern as well as example usage in a well-documented and easily-referenced way.

Reply:
Hmm yeah, that makes sense. For this PR, it would mean reverting the behavior in `Stage` that now implements `LifecycleListener` but that shouldn't be too big of a change.